core/events/message.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+import abc
+import re
+from enum import Enum
+from typing import TYPE_CHECKING, Annotated, Type, Union
+
+from pydantic import BaseModel, Discriminator, Field
+from pydantic import Tag as PydanticTag
+
+if TYPE_CHECKING:
+    from core.schemas import (
+        dfiq,
+        entity,
+        graph,
+        indicator,
+        observable,
+        tag,
+        task,
+        template,
+        user,
+    )
+
+
+class MessageType(str, Enum):
+    log = "log"
+    event = "event"
+
+
+class EventType(str, Enum):
+    new = "new"
+    update = "update"
+    delete = "delete"
+
+
+ObservableObjectTypes = Annotated[
+    "observable.ObservableTypes", Field(discriminator="type")
+]
+TaskObjectTypes = Annotated["task.TaskTypes", Field(discriminator="type")]
+EntityObjectTypes = Annotated["entity.EntityTypes", Field(discriminator="type")]
+IndicatorObjectTypes = Annotated[
+    "indicator.IndicatorTypes", Field(discriminator="type")
+]
+UserTypes = Union["user.UserSensitive", "user.User"]
+
+
+def yeti_object_discriminator(v):
+    if isinstance(v, dict):
+        return v.get("root_type", None)
+    elif isinstance(v, BaseModel):
+        return getattr(v, "root_type", None)
+    return None
+
+
+YetiObjectTypes = Annotated[
+    Union[
+        Annotated[ObservableObjectTypes, PydanticTag("observable")],
+        Annotated["dfiq.DFIQTypes", PydanticTag("dfiq")],
+        Annotated[TaskObjectTypes, PydanticTag("task")],
+        Annotated[EntityObjectTypes, PydanticTag("entity")],
+        Annotated[IndicatorObjectTypes, PydanticTag("indicator")],
+        Annotated[UserTypes, PydanticTag("user")],
+        Annotated["tag.Tag", PydanticTag("tag")],
+        Annotated["template.Template", PydanticTag("template")],
+        Annotated["graph.Relationship", PydanticTag("relationship")],
+        Annotated["graph.TagRelationship", PydanticTag("tag_relationship")],
+    ],
+    Field(discriminator=Discriminator(yeti_object_discriminator)),
+]
+
+
+class AbstractEvent(BaseModel, abc.ABC):
+    def match(self, acts_on: str) -> bool:
+        raise NotImplementedError
+
+
+class ObjectEvent(AbstractEvent):
+    type: EventType
+    yeti_object: YetiObjectTypes
+
+    def match(self, acts_on: str) -> bool:
+        return re.match(acts_on, self.event_message)
+
+    @property
+    def event_message(self) -> str:
+        event_message = f"{self.type}:{self.yeti_object.root_type}"
+        if hasattr(self.yeti_object, "type"):
+            event_message += f":{self.yeti_object.type}"
+        return event_message
+
+
+class LinkEvent(AbstractEvent):
+    type: EventType
+    source_object: YetiObjectTypes
+    target_object: YetiObjectTypes
+    relationship: "graph.Relationship"
+
+    def match(self, acts_on: str) -> bool:
+        return re.match(acts_on, self.link_source_event) or re.match(
```

Comment:
re.match will compile the regex of acts_on every time, can we somehow pre-cache this so that we only compile it once? Given how often we expect this to be called I think it's worth it (same comment for the match function in TagEvent and ObjectEvent)

Reply:
We could add this in the EventTask directly.

Reply:
fixed in f86706f1d12c71a6672e89ae3f3e63876b5f2ab4